documents/model.go

```diff
@@ -25,3 +31,132 @@ type Model interface {
 	// CreateProofs creates precise-proofs for given fields
 	CreateProofs(fields []string) (coreDoc *coredocumentpb.CoreDocument, proofs []*proofspb.Proof, err error)
 }
+
+// CoreDocumentModel contains methods which handle all interactions mutating or reading from a core document
+// Access to a core document should always go through this model
+type CoreDocumentModel struct {
+	Document *coredocumentpb.CoreDocument
+}
+
+// NewCoreDocModel returns a new CoreDocumentModel
+// Note: collaborators and salts are to be filled by the caller
+func newCoreDocModel() *CoreDocumentModel {
+	id := utils.RandomSlice(32)
+	cd := &coredocumentpb.CoreDocument{
+		DocumentIdentifier: id,
+		CurrentVersion:     id,
+		NextVersion:        utils.RandomSlice(32),
+	}
+	return &CoreDocumentModel{
+		cd,
+	}
+}
+
+// GetDocument returns the coredocument from the CoreDocumentModel
+func (m *CoreDocumentModel) GetDocument() (coredocumentpb.CoreDocument, error) {
+	cd := m.Document
+	if cd == nil {
+		return coredocumentpb.CoreDocument{}, errors.New("error getting document in CoreDocModel")
+	}
+	return *cd, nil
+}
+
+// PrepareNewVersion creates a new CoreDocumentModel with the version fields updated
+// Adds collaborators and fills salts
+// Note: new collaborators are added to the list with old collaborators.
+//TODO: this will change when collaborators are moved down to next level
+func (m *CoreDocumentModel) PrepareNewVersion(collaborators []string) (*CoreDocumentModel, error) {
+	ndm := newCoreDocModel()
+	ncd := ndm.Document
+	ocd := m.Document
+	ucs, err := fetchUniqueCollaborators(ocd.Collaborators, collaborators)
+	if err != nil {
+		return nil, errors.New("failed to decode collaborator: %v", err)
+	}
+
+	cs := ncd.Collaborators
+	for _, c := range ucs {
+		c := c
+		cs = append(cs, c[:])
+	}
+
+	ncd.Collaborators = cs
+
+	// copy read rules and roles
+	ncd.Roles = m.Document.Roles
+	ncd.ReadRules = m.Document.ReadRules
+	addCollaboratorsToReadSignRules(ncd, ucs)
+
+	err = ndm.fillSalts()
+	if err != nil {
+		return nil, err
+	}
+
+	cd, err := m.GetDocument()
```

Comment:
```suggestion
	cd  := m.Document
```
Or better let's use the vars defined at the beginning of the method, as they are pointers.